experimental/casemapping/src/internals.rs

```diff
@@ -0,0 +1,1113 @@
+            if c == '\u{cc}' {
+                return Some(FullMappingResult::String(Self::I_DOT_GRAVE));
+            } else if c == '\u{cd}' {
+                return Some(FullMappingResult::String(Self::I_DOT_ACUTE));
+            } else if c == '\u{128}' {
+                return Some(FullMappingResult::String(Self::I_DOT_TILDE));
+            }
+        }
+
+        if locale == CaseMapLocale::Turkish {
+            if c == '\u{130}' {
+                // I and i-dotless; I-dot and i are case pairs in Turkish and Azeri
+                return Some(FullMappingResult::CodePoint('i'));
+            } else if c == '\u{307}' && context.preceded_by_capital_i(self) {
+                // When lowercasing, remove dot_above in the sequence I + dot_above,
+                // which will turn into i. This matches the behaviour of the
+                // canonically equivalent I-dot_above.
+                return Some(FullMappingResult::Remove);
+            } else if c == 'I' && !context.followed_by_dot_above(self) {
+                // When lowercasing, unless an I is before a dot_above, it turns
+                // into a dotless i.
+                return Some(FullMappingResult::CodePoint('\u{131}'));
+            }
+        }
+
+        if c == '\u{130}' {
+            // Preserve canonical equivalence for I with dot. Turkic is handled above.
+            return Some(FullMappingResult::String(Self::I_DOT));
+        }
+
+        if c == '\u{3a3}'
+            && context.preceded_by_cased_letter(self)
+            && !context.followed_by_cased_letter(self)
+        {
+            // Greek capital sigman maps depending on surrounding cased letters.
+            return Some(FullMappingResult::CodePoint('\u{3c2}'));
+        }
+
+        // No relevant special case mapping. Use a normal mapping.
+        None
+    }
+
+    fn full_upper_or_title_special_case(
+        &self,
+        c: char,
+        context: ContextIterator,
+        locale: CaseMapLocale,
+        _is_title: bool,
+    ) -> Option<FullMappingResult> {
+        if locale == CaseMapLocale::Turkish && c == 'i' {
+            // In Turkic languages, i turns into a dotted capital I.
+            return Some(FullMappingResult::CodePoint('\u{130}'));
+        }
+        if locale == CaseMapLocale::Lithuanian
+            && c == '\u{307}'
+            && context.preceded_by_soft_dotted(self)
+        {
+            // Lithuanian retains the dot in a lowercase i when followed by accents.
+            // Remove dot_above after i with upper or titlecase.
+            return Some(FullMappingResult::Remove);
+        }
+        // TODO: ICU4C has a non-standard extension for Armenian ligature ech-yiwn.
+        // Should we also implement it?
+        // if c == '\u{587}' {
+        //     return match (locale, is_title) {
+        // 	(CaseMapLocale::Armenian, false) => Some(FullMappingResult::String("ԵՎ")),
+        // 	(CaseMapLocale::Armenian, true) => Some(FullMappingResult::String("Եվ")),
```

Comment:
question: would one of these need `CaseMapLocale::WesternArmenian`?

Reply:
Good question, I'll look into it.